Xamarin.Forms.Core/Binding.cs

```diff
@@ -114,10 +114,15 @@ internal override void Apply(bool fromTarget)
 			_expression.Apply(fromTarget);
 		}
 
-		internal override void Apply(object newContext, BindableObject bindObj, BindableProperty targetProperty)
+		internal override void Apply(object newContext, BindableObject bindObj, BindableProperty targetProperty, bool fromBindingContextChanged = false)
 		{
 			object src = _source;
-			base.Apply(src ?? newContext, bindObj, targetProperty);
+			var isApplied = IsApplied;
+
+			base.Apply(src ?? newContext, bindObj, targetProperty, fromBindingContextChanged: fromBindingContextChanged);
+
+			if (src != null && isApplied && fromBindingContextChanged)
```

Comment:
If I remove `fromBindingContextChanged` and change this line to simply `if (src != null && isApplied)` then all 80 unit tests in `BindableObjectUnitTests.cs`, including the two new ones, all pass. Let's include a test that fails.